include/coreir/ir/typegen.h

```diff
@@ -3,36 +3,63 @@
 
 #include "fwd_declare.h"
 #include "globalvalue.h"
+#include "common.h" //This should really be moved to fwd_declare
 
 namespace CoreIR {
 
 class TypeGen : public GlobalValue {
+  std::map<Values,Type*,ValuesComp> typeCache;
   Params params;
   bool flipped;
   //TODO maybe cache the types based off the args
-  public:
+  protected : 
     TypeGen(Namespace* ns, std::string name, Params params, bool flipped=false) : GlobalValue(GVK_TypeGen,ns,name), params(params), flipped(flipped) {}
+    virtual Type* createType(Values args) = 0;
+  
+  public :
     virtual ~TypeGen() {}
-    virtual Type* createType(Context* c, Values args) = 0;
-    virtual std::string toString() const override = 0;
-    virtual void print() const override = 0;
-    Type* getType(Values genargs); //TODO change this to a functor
+    virtual bool hasType(Values args) = 0;
+    //This is the actual 
+    virtual Type* getType(Values genargs) final;
     Params getParams() const {return params;}
+    const std::map<Values,Type*,ValuesComp>& getCached() { return typeCache;}
     bool isFlipped() const { return flipped;}
+    virtual std::string toString() const override = 0;
+    virtual void print() const override {std::cout << this->toString() << std::endl;}
 };
 
 //Notice, the base class does the flipping for you in the function computeType
 class TypeGenFromFun : public TypeGen {
+  protected :
+    TypeGenFun fun;
+    TypeGenFromFun(Namespace* ns, std::string name, Params params, TypeGenFun fun, bool flipped) : TypeGen(ns,name,params,flipped), fun(fun) {}
+    Type* createType(Values genargs) override {
+      return fun(this->getContext(),genargs);
+    }
   public :
-    TypeGenFromFun(Namespace* ns, std::string name, Params params, TypeGenFun fun, bool flipped=false) : TypeGen(ns,name,params,flipped), fun(fun) {}
-    Type* createType(Context* c, Values genargs) override {
-      return fun(c,genargs);
+    bool hasType(Values genargs) override {
+      return true; //Really should check if params are val TODO
```

Comment:
I'm not sure what this new `hasType` is supposed to do, perhaps checks if the parameters a valid set of arguments that will result in a type?